File: 0-environment-setup-and-nestjs-core/src/dog/dog.controller.ts
/**
 * Controller xử lý endpoint `/dogs/spy` — demo cross-module dependency.
 * (EN: Controller handling `/dogs/spy` endpoint — cross-module dependency demo.)
 */
import {
    Controller,
    Get,
} from "@nestjs/common"
import {
    DogService,
} from "./dog.service"

@Controller("dogs")
export class DogController {
    /**
     * Inject service để controller giữ đúng vai trò orchestration.
     * (EN: Inject service so controller stays as an orchestration layer.)
     */
    constructor(private readonly dogService: DogService) {}

    /**
     * Xử lý `GET /dogs/spy` cho bài test cross-module dependency.
     * (EN: Handle `GET /dogs/spy` for cross-module dependency test.)
     */
    @Get("spy")
    spy() {
        // Không đặt logic nghiệp vụ ở controller để giữ code dễ test và maintain.
        // (EN: Keep business logic out of controller for easier testing and maintenance.)
        return this.dogService.getSpyReport()
    }
}

File: 0-frameworks-in-backend/backend/1-java/src/main/java/com/example/dog/DogController.java
package com.example.dog;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * REST controller for the {@code /dogs/*} endpoints.
 *
 * <p>Demonstrates cross-package (cross-module) dependency injection: Spring
 * supplies {@link DogService}, which in turn receives {@code CatService} —
 * all without any {@code new} keyword in application code.
 */
@RestController
public class DogController {

    /** DogService injected by Spring; controller stays as a thin routing layer. */
    private final DogService dogService;

    /**
     * Constructor injection — Spring resolves {@link DogService} from its context.
     *
     * @param dogService the singleton {@link DogService} bean managed by Spring
     */
    public DogController(DogService dogService) {
        this.dogService = dogService;
    }

    /**
     * Handle {@code GET /dogs/spy} — proves DogService can call CatService across
     * the package boundary via injection.
     *
     * @return {@link DogService.SpyReport} with mission, dependency, and status fields
     */
    @GetMapping("/dogs/spy")
    public DogService.SpyReport spy() {
        // Delegate entirely to service; controller has no business logic.
        return dogService.spyReport();
    }

    /**
     * Handle {@code GET /dogs/cats-via-di} — proves the injected CatService is the
     * same singleton instance shared with CatController.
     *
     * @return {@link DogService.CatsView} containing dog name and borrowed cat list
     */
    @GetMapping("/dogs/cats-via-di")
    public DogService.CatsView catsViaDI() {
        return dogService.catsViaDI();
    }
}
